extensions/cache/deployment/src/main/java/io/quarkus/cache/deployment/CacheMethodValidator.java

```diff
@@ -0,0 +1,122 @@
+package io.quarkus.cache.deployment;
+
+import static io.quarkus.cache.deployment.CacheDeploymentConstants.CACHE_INVALIDATE;
+import static io.quarkus.cache.deployment.CacheDeploymentConstants.CACHE_INVALIDATE_ALL;
+import static io.quarkus.cache.deployment.CacheDeploymentConstants.CACHE_LOAD;
+import static io.quarkus.cache.deployment.CacheDeploymentConstants.CACHE_STORE;
+import static io.quarkus.cache.deployment.CacheDeploymentConstants.NON_BLOCKING_RETURN_TYPES;
+
+import java.util.List;
+import java.util.Optional;
+
+import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.DotName;
+import org.jboss.jandex.MethodInfo;
+import org.jboss.jandex.Type;
+
+import io.quarkus.arc.processor.AnnotationStore;
+import io.quarkus.arc.processor.BeanInfo;
+import io.quarkus.cache.deployment.exception.MultipleCacheAnnotationsException;
+import io.quarkus.cache.deployment.exception.NoParameterException;
+import io.quarkus.cache.deployment.exception.NonBlockingReturnTypeException;
+import io.quarkus.cache.deployment.exception.VoidReturnTypeException;
+
+public class CacheMethodValidator {
+
+    public static void validateAnnotations(AnnotationStore annotationStore, BeanInfo bean, MethodInfo method,
+            List<Throwable> throwables) {
+
+        int methodAnnotationsCount = 0;
+        if (validateCacheLoadAnnotation(annotationStore, bean, method, throwables)) {
+            methodAnnotationsCount++;
+        }
+        if (validateCacheStoreAnnotation(annotationStore, bean, method, throwables)) {
```

Comment:
I've tried to keep the annotations simple, short and trying to represent their main use case:

From a user perspective, the most important aspect of `@CacheLoad` method is that if the result of the method invocation has been cached, it will return from the cache, so it'll be loaded from the cache. Let's call that the happy path. So, I named the annotation based on the happy path.

The secondary aspect, not-so-happy path, is that if the cache is not populated, the method is called and the result is cached, but I see this as a side effect and not the most important side of the method.

It's true that you need to be aware of both paths, particularly since the not-so-happy path is potentially more expensive one in terms of IO and contention.

The important thing here is that we're taking the responsibility to cache things from the user.  That's why they don't need to worry about `@CacheStore` for actually caching data. The user needs to tell us how to retrieve things from the data source, and how to update such data and how to remove here. 

`Maybe `@CacheStore` should be really called `@CacheUpdate`. I'd still leave `@CacheLoad` as is.

Reply:
I don't think `@CacheUpdate` would be a good name because this annotation can be used to insert a new cache entry and therefore not update anything in the cache. `@CacheStore` seems a better choice.

Reply:
Hm, I think that `@CacheResult` would be probably more comprehensible than `@CacheLoad`? I know that `javax.cache` is not very popular among our engineers but the `@CacheResult` annotation name is probably one of the best parts of this API ;-).

Reply:
I think `@CacheResult` could be just as valid as `@CacheLoad` since a method with this annotation does either of these two actions. It all depends which side of the coin you look at, e.g.

a) If the cache does not contain the cached entry, the method is called and the **result is cached**.
b) If the cache contains the cached entry, the result is **loaded from the cache**.

If I had to spell it out, I'd say the annotation should be called `@LoadFromCacheOrCacheResult` ;)

Might need to do a poll on this, but it'd be great to hear other suggestions.